File: dailyQuestion/2021/2021-03/03-06/golang/solution.go
package main

import "fmt"

func main() {
	nums := []int{1, 2, 1, 7, 6, 4, 9, 8,1}
	res := nextGreaterElements(nums)

	fmt.Println(res)

}

func nextGreaterElements(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	for i := 0; i < n; i++ {
		res[i] = -1
	}
	stack := []int{}
	for i := 0; i < 2*n-1; i++ {
		for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[i%n] {
			res[stack[len(stack)-1]] = nums[i%n]
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i%n)
	}
	return res
}

File: dailyQuestion/2021/2021-03/03-06/python/solution.py
class Solution:
    def nextGreaterElements(self, nums: [int]) -> [int]:
        n = len(nums)
        res = [-1] * n
        stack = []
        for i in range(2 * n -1) :
            while len(stack) > 0 and nums[stack[len(stack) -1]] < nums[i %n] :
                res[stack[len(stack) -1]] = nums[i %n]
                stack.pop()
            stack.append( i % n)
        return res

if __name__ == "__main__" :
    s = Solution()
    nums = [1,2,1,7,6,4,9,8,1]
    res = s.nextGreaterElements(nums)
    print(res)